### PR TITLE
Feature/copy command to clipboard

### PR DESCRIPTION
**Pull request by Sagardevkota**
## Add Copy Button for "npx nx ...." Command

- When navigating the Angular Challenges website's documentation, I often find myself manually copying the "npx nx ..." command from the Note section. To enhance user experience, I propose the addition of a convenient copy button. This button will automatically copy the command to the clipboard, allowing users to effortlessly paste it into their terminal for execution.
- New changes would look like following screenshot:

image

### CODE REVIEW

**Comment by tomalaforge** [comment]
This is such a nice idea and great improvement. 😍

One small change, can you create a component so we can maybe use it in other places.

**Comment by tomalaforge** [comment]
@all-contributors add @Sagardevkota for code

**Comment by Sagardevkota** [comment]
> This is such a nice idea and great improvement. 😍
> 
> One small change, can you create a component so we can maybe use it in other places. And can you do something like in Github, with a checkmark when the line is copied ?

Thanks for your valuable feedback.. I have created a ClipboardCopy.astro component & moved all the code there.. i am fairly new to astro.. You can point if there is a better way :) .. I think the Github is using setTimeout but i have opted to use mouseleave event.. i can change to whatever you prefer.. 

Thanks Again

**Comment by tomalaforge** [comment]
Nice, better with the component. 

I think I prefer the setTimeout. With mouseLeave, the UX is weird. We don't really understand if the click worked or not. The button is very small, so we move outside of it really fast and the icon reset too fast. A timeout of 3/4s will feel more natural

File: docs/src/components/ClipboardCopy.astro [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  let timeout;
+  commandCopyBtn.addEventListener('click', () => {
+    navigator.clipboard.writeText(copyText);
+    tooltipText.innerHTML = 'Copied';
+    copiedCheckMark.style.display = 'inline-block';
+    commandCopyBtn.style.display = 'none';
+    //clear previous timeout
+    clearTimeout(timeout);
+    timeout = window.setTimeout(()=>{
```
Comment (tomalaforge) [inline docs/src/components/ClipboardCopy.astro:51]: You don't need to use window.setTimeout, 
you can use setTimeout and don't need to store it and clear it.

File: docs/src/components/ClipboardCopy.astro [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    copiedCheckMark.style.display = 'none';
+    commandCopyBtn.style.display = 'inline-block';
+    tooltipText.innerHTML = 'Copy';
+  }, 750)
```
Comment (tomalaforge) [inline docs/src/components/ClipboardCopy.astro:55]: 2000 would be a bit better

File: docs/src/components/ClipboardCopy.astro [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  });
+</script>
+
+<style>
+  .copy-button {
+    fill: var(--sl-color-accent-high);
```
Comment (tomalaforge) [inline docs/src/components/ClipboardCopy.astro:61]: I think it's better if it's gray, and red on hover

File: docs/src/components/ClipboardCopy.astro [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    margin-left: 0.5rem;
+    cursor: pointer;
+  }
+
+  .copied-checkmark {
+    fill: var(--sl-color-accent-high);
```
Comment (tomalaforge) [inline docs/src/components/ClipboardCopy.astro:67]: like github, green is a good color to show that it worked

File: docs/src/components/ClipboardCopy.astro [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  const commandCopyBtn = document.querySelector('.copy-button');
+  const copiedCheckMark = document.querySelector('.copied-checkmark');
+  const tooltipText = document.querySelector('.tooltipText');
+
+  let timeout;
+  commandCopyBtn.addEventListener('click', () => {
+    navigator.clipboard.writeText(copyText);
+    tooltipText.innerHTML = 'Copied';
+    copiedCheckMark.style.display = 'inline-block';
+    commandCopyBtn.style.display = 'none';
```
Comment (tomalaforge) [inline docs/src/components/ClipboardCopy.astro:48]: you can imitate github and show the Copied tooltip when clicked. 👍

**Comment by tomalaforge** [comment]
Very nice, I added a few last comment to make it perfect.

File: docs/src/components/ClipboardCopy.astro (hunk 5 above)
Comment (Sagardevkota) [reply docs/src/components/ClipboardCopy.astro:47]: > you can imitate github and show the Copied tooltip when clicked. 👍

Thanks for your feedback. I have changed all the color based on your recommendation & tried to imitate github :smiley

**Comment by jdegand** [comment]
Uncaught (in promise) TypeError: Failed to construct 'URL': Invalid URL error in console. 

This error is not present on main Angular Challenges app.   

Styling issues : 

Note and Reminder are gone.  

.starlight-aside--note in custom-css file -> need to remove background:transparent.    

The copied text isn't displaying.

Need remove setting the color on .tooltipText (line 82).

Probably best to create a command component with the copy button and add the component where all commands are used in the whole docs ?

**Comment by Sagardevkota** [comment]
> Uncaught (in promise) TypeError: Failed to construct 'URL': Invalid URL error in console.
> 
> This error is not present on main Angular Challenges app.
> 
> Styling issues :
> 
> Note and Reminder are gone.
> 
> .starlight-aside--note in custom-css file -> need to remove background:transparent.
> 
> The copied text isn't displaying.
> 
> Need remove setting the color on .tooltipText (line 82).
> 
> Probably best to create a command component with the copy button and add the component where all commands are used in the whole docs ?

I don't have any errors/issues in my local setup.. I have changed tooltip background to red to fix white bg & color on light theme

**Comment by jdegand** [comment]
Maybe error is caused by problem with vercel deployment.

**Comment by tomalaforge** [comment]
That's perfect, thanks a lot @Sagardevkota

File: docs/src/components/ClipboardCopy.astro [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        tooltipText.style.display = 'none';
+  }, 2000)
+  });
+</script>
+
+<style>
+  .copy-button {
+    margin-left: 0.5rem;
+    cursor: pointer;
+  }
```
Comment (tomalaforge) [inline docs/src/components/ClipboardCopy.astro:61]: just for information: you can do 
```css
.copy-button {
margin-left: 0.5rem;
    cursor: pointer;
    
    &:hover {
fill: var(--sl-color-accent-high);.}
}
```

**Review by tomalaforge — APPROVED** [review]

**Pull request merged**